### PR TITLE
Fixes #20 - issue with inproperly merging conflicting rules

### LINKED ISSUE #20: Wrong order when partially merging rules with specializing records
There is a (seems to be) wrong behaviour when mergin two rules of this type:

```
.foo {
    margin: 0;
    margin-left: 30px;
} 

.bar {
    margin: 0;
    margin-right: 30px;
}
```

Which is converted to:

```
.foo {
    margin-left: 30px; /* will be overwriten by next rule */
}

.foo,.bar {
    margin: 0;
}

.bar {
    margin-right: 30px;
}
```
Left margin here will be lost.

Are there any reasons to keep merged rule after first one? Pushing merged rule before first and second looks like an easiest solution which will solve many order problems but I don't know css enough to predict what can go wrong.

### PR DESCRIPTION
**Pull request by 11bit**
I've added a check that prevents merging of rules that overwrite themselves (issue  #20). Looks a bit complicated but I don't see easier way. What do you think?

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -67,14 +77,29 @@ function partialMerge (first, second) {
...
+            let prop = decl.split(':')[0];
+            let base = prop.split('-')[0];
+            let canMove = difference.every(d => d.split(':')[0] !== base);
+            if (intersects && canMove && !hasConflicts(prop, willNotMove)) {
+                willMove.push(decl);
+            } else {
+                willNotMove.push(prop);
+            }
+        });
+        return willMove;
```
Comment (ben-eb) [inline src/index.js]: This method could be written like this instead:

```js
let willNotMove = [];
return decls.reduce((willMove, decl) => {
    let intersects = ~intersection.indexOf(decl);
    let prop = decl.split(':')[0];
    let base = prop.split('-')[0];
    let canMove = difference.every(d => d.split(':')[0] !== base);
    if (intersects && canMove && !hasConflicts(prop, willNotMove)) {
        willMove.push(decl);
    } else {
        willNotMove.push(prop);
    }
    return willMove;
}, []);
```

File: src/index.js [hunk 2]
```diff
@@ -50,6 +50,16 @@ function ruleLength (...rules) {
...
 }
 
+function isConflictingProp(propA, probB) {
+    let a = propA.split('-');
+    let b = probB.split('-');
+    return a.length !== b.length && a[0] === b[0];
+}
+
+function hasConflicts(declProp, notMoved) {
+    return notMoved.some(prop => isConflictingProp(prop, declProp))
```
Comment (ben-eb) [inline src/index.js]: Missing semicolon. It's my fault, should have migrated this module over to ESLint but haven't found the time yet. :smile:

**Comment by ben-eb** [comment]
Really awesome work, thanks! Could you check that this resolves the test cases posted in these issues too please?

https://github.com/ben-eb/postcss-merge-rules/issues/13
https://github.com/ben-eb/postcss-merge-rules/issues/14
https://github.com/ben-eb/postcss-merge-rules/issues/21

https://github.com/ben-eb/cssnano/issues/154
https://github.com/ben-eb/cssnano/issues/160

Would be really great to finally close this bug. :+1:

**Comment by 11bit** [comment]
I have tested most of the minimal examples in these issues. Looks like it fixes all of them (added one modification for #14)
I just realised that my conflicts testing algo will have false positives on rules with vendor prefixes. It won't allow to merge something like:
`.a {-webkit-transform: ..., -webkit-border-radius: ...}, .b {-webkit-transform: ...}`
I will think about it

**Comment by ben-eb** [comment]
OK. :+1:

**Comment by 11bit** [comment]
Added a code for proper handling of vendor's prefixes

**Comment by ben-eb** [comment]
Cool. Can you patch the code according to my comments please and add those test cases? This has been an issue for a lot of our users so I would like to do it properly. :smile:

**Comment by 11bit** [comment]
Sorry, I didn't noticed your comments - just changed this.
I've also added more tests, but the cases are very similar one to each other. Don't know if it is worth it to add more.

**Comment by 11bit** [comment]
@ben-eb How else can I help with this issue?

**Comment by ben-eb** [comment]
I've been a bit busy, need to check the integration tests for cssnano and we can merge this. It'll land soon. :smile:

**Comment by 11bit** [comment]
Ok, no problem :)

**Comment by ben-eb** [comment]
Thanks again, landed in 2.0.7. 🎉

**Comment by 11bit** [comment]
cool :)

**Pull request merged**